File: Mis Prácticas/NapakalakiRuby/lib/cultist_player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "player.rb"
require_relative "cultist.rb"

class CultistPlayer < Player
  
  @@totalCultistPlayer = 0
  
  def initialize(p, c)
    super(p)
    @myCultistCard = c
    @@totalCultistPlayer += 1    
  end
  
  def get_my_cultist_card
    @myCultistCard
  end
  
  def self.get_total_cultist_players
    @@totalCultistPlayer
  end
  
  protected
  def get_combat_level
    (super + super*0.2 + 
      @myCultistCard.get_gained_levels * @@totalCultistPlayer).ceil
  end
  
  def get_oponent_level(m)
    m.get_combat_level_against_cultist_player
  end
  
  def should_convert
    false
  end
  
  def give_me_a_treasure
    @visibleTreasures[rand(@visibleTreasures.length)]
  end
  
  def can_you_give_me_a_treasure
    return @visibletreasures.any? 
  end
  
  def remove_treasure(t)
    @visibletreasures.delete(t)
  end
  
  public  
  def to_s
    super + ", Cultist: si"
  end
    
end

File: Napakalaki/src/napakalaki/CultistPlayer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;

import java.util.Random;

/**
 *
 * @author fran
 */
public class CultistPlayer extends Player{
    
    // Variables de clase
    private static int totalCultistPlayer = 0;
    
    // Variables de instancia
    private Cultist myCultistCard;
    
    // Constructores
    public CultistPlayer(Player p, Cultist c){
        
        super(p);
        setMyCultistCard(c);
        
        totalCultistPlayer++;
        
    }
    
    // Getters
    public Cultist getMyCultistCard(){
        return myCultistCard;
    }
    
    public static int getTotalCultistPlayers(){
        return totalCultistPlayer;
    }
    
    // Setters
    private void setMyCultistCard(Cultist myCultistCard){
        this.myCultistCard = myCultistCard;
    }
    
    // Utils
    
    @Override
    protected int getCombatLevel(){        
        return (int)(super.getCombatLevel() + (super.getCombatLevel()*0.2) + 
                (myCultistCard.getGainedLevels() * totalCultistPlayer));
    }
    
    @Override
    protected int getOponentLevel(Monster m){
        return m.getCombatLevelAgainstCultistPlayer();
    }
    
    @Override
    protected boolean shouldConvert(){
        return false;
    }
        
    @Override
    protected Treasure giveMeATreasure(){
        return super.getVisibleTreasures().get(new Random().nextInt(super.getVisibleTreasures().size()));
    }
    
    
    @Override
    protected boolean canYouGiveMeATreasure(){
        return !super.getVisibleTreasures().isEmpty();
    }
    
    @Override
    protected void removeTreasure(Treasure treasure){
        super.getVisibleTreasures().remove(treasure);
    }
    
    @Override
    public boolean isCultist(){
        return true;
    }
            
}
